Add guide for adding left panels to apps (#8)
denniskigen/o3-docs: pages/docs/recipes/adding-a-left-panel.mdx

  "backendDependencies": {
    "fhir2": "^1.2.0",
    "webservices.rest": "^2.24.0"
  },
  "pages": [
    {
      "component": "root",
      "route": "bed-management"
    }
  ]
}
```

This page definition tells O3 to render the `Root` component when you navigate to the `bed-management` route.

### Step 3: Add links to the left panel

Next, we'll want to add the two links that we mentioned earlier to the left panel:

- A `Summary` link for the landing screen.
- A `Ward allocation` link for the ward allocation screen.

To do this, we'll need to create two extensions that render the links. We'll then add those extensions to the `extensions` array of our `routes.json` file. The named exports of those extensions will be used as the `component` property of the extensions, and look like the following:

```ts copy
// src/index.ts
export const summaryLeftPanelLink = getSyncLifecycle(
  createLeftPanelLink({
    name: "summary",
    title: t("summary", "Summary"),
  }),
  options
);

export const wardAllocationLeftPanelLink = getSyncLifecycle(
  createLeftPanelLink({
    name: "ward-allocation",
    title: t("wardAllocation", "Ward Allocation"),
  }),
  options
);

> ```

Reviewer: You will have to pass the `moduleName` in the useTranslation hook in the code.

> ```

Reviewer 2: @vasharma05 Note that the code here _isn't_ using the `createDashboardLink()` function from the patient chart, so the requirements here are a bit different. Also, please note that you _only_ need comments with the `t` function if:

1. The translation is needed at runtime but otherwise doesn't appear in code
2. The key to be translated cannot be inferred from the code

Neither of those two apply here.

import Image from "next/image";
import { Steps } from "nextra-theme-docs";

# Adding a left panel

> The [left panel](https://zeroheight.com/23a080e38/p/796c6a-left-panel) provides navigation within an App in O3. It is closely based on Carbon's [UI shell left panel](https://carbondesignsystem.com/components/UI-shell-left-panel/usage/), and is positioned below the header, fixed to the left edge of the page. It is possible to add a left panel to any app in O3 if you need to leverage its navigation capabilities.

Reviewer 2: ```suggestion
The [left panel](https://zeroheight.com/23a080e38/p/796c6a-left-panel) provides navigation within an App in O3. It is based on Carbon's [UI shell left panel](https://carbondesignsystem.com/components/UI-shell-left-panel/usage/), and is positioned below the header, fixed to the left edge of the page. It is possible to add a left panel to any page in O3 if you need to leverage its navigation capabilities.
```

- A link named `Summary` which is where you first land after visiting the `/bed-management/summary` route.
- A link named `Ward allocation` which maps to the `/bed-management/administration` route.

Below is a screenshot of how the landing page looks like:

<br />
<Image
  src="/left-panel-home.png"
  alt="Screenshot of the bed management app landing page showing the left panel"
  width={1000}
  height={1000}
/>
<br />

To achieve this, we're going to follow the steps below:

<Steps>
### Step 1: Set up the left panel

We'll start by setting up the left panel. To do so, we'll create a directory named `left-panel` and add a component that renders the left panel.

```tsx copy
// src/left-panel/left-panel.component.tsx
import React, { useEffect } from "react";
import {
  attach,
  detach,
  ExtensionSlot,
  isDesktop,
  useLayoutType,
} from "@openmrs/esm-framework";
import { SideNav } from "@carbon/react";
import styles from "./left-panel.scss";

const LeftPanel: React.FC = () => {
  const layout = useLayoutType();

  useEffect(() => {
    attach("nav-menu-slot", "bed-management-left-panel");
    return () => detach("nav-menu-slot", "bed-management-left-panel");
  }, []);

  return (

>     isDesktop(layout) && (

Reviewer 2: Should the example include how we handle this on tablet? Or is this built in?

Author: It's built-in.

  useEffect(() => {
    setLeftNav({
      name: "bed-management-left-panel-slot",
      basePath: spaBasePath,
    });
    return () => unsetLeftNav("bed-management-left-panel-slot");
  }, [spaBasePath]);

  return (
    <BrowserRouter basename={`${window.getOpenmrsSpaBase()}bed-management`}>
      <LeftPanel />
      <main className={styles.container}>
        <Routes>
          <Route path="/summary" element={<Home />} />
          <Route path="/location/:location" element={<WardWithBeds />} />
          <Route path="/administration" element={<BedAdministrationTable />} />
        </Routes>
      </main>
    </BrowserRouter>
  );
};

export default Root;
```

Key things to note here are:

- We're registering our slot with the `LeftNav` store when the component mounts, setting the provided spaBasePath (`/openmrs/spa`) as the basePath. We're also unregistering the slot from the LeftNav store when the component unmounts. The [LeftNav store](https://github.com/openmrs/openmrs-esm-core/blob/da401641b4b1895e593e9bcea73af8f8dbd505e5/packages/framework/esm-styleguide/src/left-nav/index.tsx#L18) is essentially a Zustand store that tracks slot names and their associated `basePath`s.

> - We're adding the `LeftPanel` component at the top of the `<BrowserRouter>` tree, just outside the routes declarations. This means that the `LeftPanel` gets rendered in all the pages defined in this app.

Reviewer 2: I'm being a bit pedantic on this point, but I think we need to consistently use "app" (collection of related extensions, pages, and functionality), "page" (a MF that is registered as a page), etc.

```suggestion
- We're adding the `LeftPanel` component at the top of the `<BrowserRouter>` tree, just outside the routes declarations. This means that the `LeftPanel` gets rendered in all the screens that use this page.
```

- We're making a small tweak to ensure that both the Summary page and the Ward detail page use the same left panel link. When the last segment of the URL is a UUID (which is the case when you click on a ward card on the landing page), we're setting the URL segment to `summary`. This ensures that both the landing and detail pages use the same link. Your mileage will likely vary when it comes to setting up your navigation links, and this is just one approach for solving that problem.
- Finally, we're wrapping the `LinkExtension` component in a `BrowserRouter` component which provides routing functionality.

Next, we'll modify our `index.ts` file to add named exports for extensions in the left panel using the `createLeftPanelLink` higher-order function from above:

```ts
// src/index.ts
export const summaryLeftPanelLink = getSyncLifecycle(
  createLeftPanelLink({
    name: "summary",
    title: "Summary",
  }),
  options
);

export const adminLeftPanelLink = getSyncLifecycle(
  createLeftPanelLink({
    name: "administration",

>     title: "Ward Allocation",

Reviewer 2: ```suggestion
    title: t("Ward Allocation"),
```
